payment/watcher.py

```diff
@@ -1,80 +1,158 @@
+        self.enqueuer = enqueuer
+        self.webhook = webhook
+        self.bc_manager = bc_manager
+        self.watched_addresses: Dict[str, Set[str]] = {}  # Map address:orders
+        self.task: asyncio.Task = None
+        self.executor = ProcessPoolExecutor()
+        self.loop = asyncio.get_running_loop()
+        self._expiration_tasks: Dict[str, asyncio.TimerHandle] = {}
+
+    def watch(self, address: str, order: str):
+        """Start watching an order for an address"""
+        if address in self.watched_addresses:
+            self.watched_addresses[address].add(order)
+        else:
+            self.watched_addresses[address] = {order}
+        self._expiration_tasks[order] = self.loop.call_later(ORDER_EXPIRY, self.remove, address, order)
+        log.debug(f'Added order: {order} to {address} watch list')
+
+    def remove(self, address, order):
+        """Remove an order from an address, and the address from the list if possible"""
+        self.watched_addresses[address].remove(order)
+        log.debug(f'Removed {order} from {address} watch list')
+        if len(self.watched_addresses[address]) == 0:
+            del self.watched_addresses[address]
+            log.debug(f'Removed {address} from watch list')
+
+        self._expiration_tasks[order].cancel()
+
+    def start(self):
+        log.info(f'Starting payment watcher')
+        self.task = asyncio.create_task(self.worker())
+
+    def stop(self):
+        log.info(f'Stopping payment watcher')
+        if self.task is not None:
+            self.task.cancel()
+        # This only happens when closing the server,
+        # so its fine to block until we close the processes
+        self.executor.shutdown(wait=True)
 
-stop_event = threading.Event()
-log = get_log()
-SEC_BETWEEN_RUNS = 1
+    async def get_last_ledger_sequence(self):
+        seq = await self.bc_manager.get_last_ledger_seq()
+        log.info('got ledger seq from horizon', seq=seq)
+        return seq
 
+    def on_payment(self, address, order_id, tx: SimplifiedTransaction):
+        """handle a new payment from/to an address."""
+        payment = Payment.from_blockchain(tx)
+        log.info(f'got payment for order {order_id} for address {address}', payment=payment)
+        statsd.inc_count('payment_observed',
+                         payment.amount,
+                         tags=['app_id:%s' % payment.app_id,
+                               'address:%s' % address])
 
-def get_last_cursor():
-    cursor = CursorManager.get()
-    if not cursor:
-        @retry(5, 0.2)
-        def get_from_horizon():
-            return Blockchain.get_last_cursor()
-        cursor = get_from_horizon()
-        log.info('got payment cursor from horizon', cursor=cursor)
-        CursorManager.save(cursor)
+        self.enqueuer.enqueue_payment_callback(self.webhook, address, payment)
+        self.remove(address, order_id)
```

Comment:
I think removing the watcher should happen only after payment callback called successfully (got 200 from the marketplace), o.w. we're not guaranteeing payment was announced to marketplace (though generally as it's on memory, in any case, we can't guarantee that in case watcher had died)